Added entropy to stats module

czgdp1807 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References
The following papers and links were used and will be used to make the necessary changes:-
1. http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf
2. https://www.crmarsh.com/static/pdf/Charles_Marsh_Continuous_Entropy.pdf
3. https://en.wikipedia.org/wiki/Entropy_(information_theory)
4. https://en.wikipedia.org/wiki/Normal_distribution#Maximum_entropy

#### Brief description of what is fixed or changed
Entropy for univariate continuous distributions has been added.

#### Other comments
Existence of methods to calculate entropy were unkown to me until the making of this PR.
I have included the `entropy` after studying the module so far.  

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* stats
  * Addition of entropy to stats module
<!-- END RELEASE NOTES -->

gdp1807 committed:
Added entropy to stats module

Entropy for univariate continuous distributions has been added.

gdp1807 committed:
Update crv.py

gdp1807 committed:
Update test_rv.py

czgdp1807 commented:
Yeah, apologies, I just wrote that for debugging purposes. 😅

gdp1807 committed:
Merge https://github.com/czgdp1807/sympy into entropy

gdp1807 committed:
Added entropy for discrete random variables

gdp1807 committed:
Added entropy for finite random variables

czgdp1807 commented:
@ethankward Kindly review the changes I have added for discrete random variables and finite random variables.

gdp1807 committed:
Removed entropy from Poisson distribution

czgdp1807 commented on sympy/stats/drv_types.py:
```diff
@@ -283,6 +286,10 @@ def _characteristic_function(self, t):
 
+    def _entropy(self, t):
+        # https://pure.tue.nl/ws/portalfiles/portal/1959440/Metis199989.pdf
+        return log(2*pi*exp(1)*t)/2 - 1/(12*t) + O(t**-2)
```
@ethankward Removed `_entropy` from `PoissonDistribution`.
Thanks for the feedback.

gdp1807 committed:
Added _entropy method to continuous distributions

_entropy method was add to ArcsinDistribution, BetaDistribution,
CauchyDistribution, ChiDistribution, ChiSquaredDistribution,
ExponentialDistribution, FDistributionDistribution, FrechetDistribution,
GammaDistribution, GammaInverseDistribution, GumbelDistribution,
and KumaraswamyDistribution. In addition, check method was added to
BeniniDistribution.

gdp1807 committed:
Added entropy to remaining continuous distributions

Entropy has been added to all single variable continuous distributions.
A few test have been modified.

gdp1807 committed:
Updated test_rv.py

czgdp1807 commented:
@ethankward Can you please tell me should `entropy` return a `Lambda` or a constant expression as result?
I was thinking of the latter one since entropy is a definite integral. Waiting for your response.

czgdp1807 commented:
@jksuom Please help me with the query in the above comment.

jksuom commented:
I think that entropy is a constant, not a Lambda. It may contain parameters if the probability distribution depends on parameters.

czgdp1807 commented:
So I think returning a constant is a better option. We can leave on the user to use Lambda wherever required. What do you think?

jksuom commented:
Lambda is used to return a function, but entropy is not a function.

czgdp1807 commented:
Thanks for the response. I will make the edits.

gdp1807 committed:
Added docs for entropy and trivial changes

The entropy will now return a constant.
Documentation has been added to entropy in
sympy.stats.rv

gdp1807 committed:
Allowed conditions for frv entropy

Conditional finite distributions can now
be passed in entropy. Docs of entropy
have also been rectified.

czgdp1807 commented:
@jksuom @Upabjojr Please review the final changes I have made to this PR. No more additions are to be made from my side.

jksuom commented:
It seems to me that entropy is the expectation of `-log(pdf(x))`. Would it be possible to use `compute_expectation` that should work for all distributions, continuous and discrete alike?

czgdp1807 commented:
Yeah, @jksuom entropy is the expectation of `-log(pdf(x))` but when using `E(-log(density(X)(x)))` works well for continuous random variables but gives `Piecewise` functions for discrete and finite random variables. So I thought of making separate methods for different types. Though if you still say, that we should use `compute_expectation` for re-usability of code then I will make the necessary changes.
Your response awaited.

jksuom commented:
For discrete variables, I would have expected a sum, either unevaluated or evaluated. Maybe there should not be both `X` and `x` in the expression.

gdp1807 committed:
Merge branch 'master' into entropy

czgdp1807 commented:
@jksuom I gave a thought to your point and came up with the following idea,
1. We can use `compute_expectation` for calculating the entropy of continuous random variable since it works out pretty well.
2. For discrete random variables we can go for the definition of entropy given [here](http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf).
What do you say?

jksuom commented:
There are many precomputed entropies but no examples of more general cases like random variables restricted to an interval.

czgdp1807 commented:
@jksuom I cannot understand your point in the latest comment. Can you please elaborate more?

jksuom commented:
In crv.py, there are 3 methods named `compute_entropy` but only one test for continuous variables (`X`), and even that is precomputed. There seem to be no tests for the `compute_entropy` methods.

czgdp1807 committed:
Generalized test cases

czgdp1807 committed:
Updated test_rv.py

czgdp1807 committed:
Merge branch 'entropy' of https://github.com/czgdp1807/sympy into entropy

czgdp1807 commented:
> @jksuom I gave a thought to your point and came up with the following idea,
> 
> 1. We can use `compute_expectation` for calculating the entropy of continuous random variables since it works out pretty well.
> 2. For discrete random variables we can go for the definition of entropy given [here](http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf).
>    What do you say?

@jksuom Any views on this?

jksuom commented:
I believe that the entropy in the discrete case is well established (the sum of `-log(p_i)*p_i`). It behaves as expected: It is non-negative and greater uncertainty implies greater entropy.

The continuous case is more difficult to define consistently. The probability distribution function pdf is scale dependent. It can take large values, which make the entropy negative. Apparently, it need not increase with greater uncertainty.
```
>>> X = Normal('X', 0, 1)
>>> H(X).n(2)
1.4
>>> H(X, X > 0).n(2)
1.5
```
I would expect that there would be less uncertainty when the domain is restricted to `X > 0`. However, the entropy seems to grow in the example. Is there a bug somewhere?

oscarbenjamin commented:
The first should be:
```
In [1]: S(1)/2 * log(2*pi*E)                                                                                                                                  
Out[1]: 
log(2⋅ℯ⋅π)
──────────
    2     

In [2]: _.evalf()                                                                                                                                             
Out[2]: 1.41893853320467
```
So looks correct (for the 2 digits you showed).

The definition of differential entropy (assuming that's what this is calculating) would suggest that the second example should be the same minus `log(2)` since
```
    H0 = - Integral(f(x) * log(f(x)), (x, -oo, oo)) 
    H1 = - Integral(2*f(x) * log(2*f(x)), (x, 0, oo)) 
       = - Integral(f(x) * log(2*f(x)), (x, -oo, oo))
       = - Integral(f(x) * log(f(x)), (x, -oo, oo)) - Integral(f(x) * log(2), (x, -oo, oo))
       = H0 - log(2)
```

So it should be
```
In [3]: S(1)/2 * log(2*pi*E) - log(2)                                                                                                                         
Out[3]: 
          log(2⋅ℯ⋅π)
-log(2) + ──────────
              2     

In [4]: _.evalf()                                                                                                                                             
Out[4]: 0.725791352644727
```

czgdp1807 committed:
Fixed bugs and updated tests

czgdp1807 commented:
@jksuom @oscarbenjamin Thanks for pointing out my mistake. I forgot to add the lower_bound and upper_bound on integrals. I have updated the PR.

jksuom commented on sympy/stats/crv.py:
```diff
@@ -258,7 +258,9 @@ def compute_entropy(self, **kwargs):
-        h = integrate(-pdf * log(pdf), (x, -oo, oo))
+        lower_bound = self.set.start
+        upper_bound = self.set.end
+        h = integrate(-pdf * log(pdf), (x, lower_bound, upper_bound))
```
Would it be possible to use something like `self.expectation` or `self.compute_expectation` that would automatically set the proper integration limits?

czgdp1807 replied:
I don't think this should be done, since it will give too complex results even for simple answers. Instead, to avoid repetition of code, we can implement a method `stats_integrate` which can be used in computing, `expectation`, `entropy`, `mgf`, `cf` alike. It will automatically set the domain.

jksuom replied:
In which way would the results be more complicated?

czgdp1807 replied:
Using compute_expectation works [here](https://github.com/sympy/sympy/pull/15693/files#diff-6f227c2ec704bbb00895cfa85e7f4d03R557) and gives results as expected. However, it in the above method was giving complex and unevaluated Integrals. Thanks for the suggestion. I will update the PR and apologies for commenting without trying the `compute_expectation` as I mentioned above. :-)

jksuom commented on sympy/stats/tests/test_rv.py:
```diff
@@ -112,8 +112,8 @@ def test_H():
-    assert H(X, X < 0) == -log(2) + 1 + log(pi)
-    assert H(X, X > 0, X < 1) == -log(2) + 1 + log(pi)
+    assert H(X, X > 0) == -log(2)/2 + 1/2 + log(pi)/2
+    assert H(X, X <= 0) == -log(2)/2 + 1/2 + log(pi)/2
```
This is essentially a repetition. Perhaps something like `H(X, X > 1)` would not look too complicated.

czgdp1807 committed:
Simplified the code for entropy computation

czgdp1807 commented:
`@jksuom Please review the changes I have made on your suggestions.

jksuom commented:
> expectation gives unevaluated sums

Perhaps `.doit()` could be added in such a case.

czgdp1807 commented:
@jksuom 
1. I don't think `.doit()` will make an effect, since there is already one as shown [here](https://github.com/sympy/sympy/blob/5d45c0adc1cbcbfcf9e07fb0bf974a379826e64d/sympy/stats/rv.py#L691)
2. When I try to execute the following code, error, `TypeError: compute_expectation() got an unexpected keyword argument 'evaluate'`. Is this error need to be fixed since, finding expectation of `L > 2` is mathematically correct, so ideally it should be handled, or conditional expressions are to be left out due to some reasons. If it is so, then I will try to fix this in this PR itself.

```
from sympy import *
from sympy.stats import *
# from sympy.stats.tests.test_rv import test_H
# X = Normal('X', 0, 1)
p = symbols('p')
L = Geometric('L', p)
print(E(L, L > 2))
```

czgdp1807 commented:
@jksuom Maybe entropy can be added only to `rv_interface` using the concept of expectation, since, a distribution can be specified without using entropy. So, adding it as a feature would be more good looking than adding it to crv_types, drv_types, frv_types. I will update if you agree.

czgdp1807 committed:
simplified code to calculate entropy

czgdp1807 committed:
formatting and missing imports

czgdp1807 committed:
formatting and imports

czgdp1807 committed:
formatting

czgdp1807 commented:
@jksuom @oscarbenjamin I have simplified the code to compute entropy. Please take a look at it and provide your reviews. Read my last comment too.

czgdp1807 committed:
removed white space

czgdp1807 committed:
updated docs example

czgdp1807 commented:
@Upabjojr Please review this. The [failing job](https://travis-ci.org/sympy/sympy/jobs/516159575) is passing on my system.

oscarbenjamin commented:
The tests are not run directly on the PR but on the master branch after merging the PR in. Other changes on master mean that in combination with your PR that test fails.

Assuming you've never committed/merged anything directly to master you can test this with
1. git fetch upstream
2. git checkout master
3. git rebase upstream/master
4. git checkout -b test_merge_15693
5. git merge entropy
6. bin/test sympy/stats/tests/test_continuous_rv.py -k test_benini

czgdp1807 committed:
resolved baseline failures

czgdp1807 committed:
resolved baseline failures

czgdp1807 committed:
resolved baseline failures

czgdp1807 commented:
@jksuom Please provide reviews on [this](https://github.com/sympy/sympy/pull/15693#issuecomment-477902372) and the changes I have made to the code.

jksuom commented:
I would like to see the entropy of a discrete random variable `X` defined as the expectation of another random variable `-log(p)` where `p` is the probability distribution of `X`. Then `p` is also a random variable (with values between 0 and 1). It may be constant (as in the case of a dice) but, in general, not expressible as a function of `X`. I think that there should be a method for creating this random variable `p` from `X`. Then the definition of entropy would be easy to write down.

czgdp1807 commented:
@jksuom I have done something similar to your suggestion. Please take a look.
In [this](https://github.com/sympy/sympy/pull/15693/files#diff-f72fbde05ac7d76ac18a349b68076190R118) line, `pdf(expr)` is a random variable. In other words, it denotes, `P(X)`, where X is a random variable and that line calculates, `E[-log(P(X))]`. It's giving the desired results correctly.

czgdp1807 committed:
resolved test failures

jksuom commented:
>  this applies to any random variable, i.e., both continuous and discrete.

Is there an example or a test for a discrete variable?

czgdp1807 commented:
Yes, I have included an example in the docs and a test in the test file.

czgdp1807 commented:
@jksuom Please provide your reviews, all tests are passing.

jksuom commented:
Can you add a few more tests? For instance, a discrete variable and one where pdf is a dict (if those exist).

czgdp1807 committed:
Added more tests, provision for base

czgdp1807 committed:
removed a minor bug

czgdp1807 commented:
@jksuom I have added a test for `Geometric` random variable. `assert simplify(H(D, D > 2)) == log(2)` tests for the random variable with pdf as `dict`.

czgdp1807 commented:
@jksuom Any more changes needed to be made? Please let me know.

jksuom commented on sympy/stats/tests/test_rv.py:
```diff
@@ -107,6 +107,14 @@ def test_IndependentProductPSpace():
+    X = Normal('X', 0, 1)
+    D = Die('D', sides = 4)
+    G = Geometric('G', 0.5)
+    assert str(H(X, X > 0)) == '-log(2)/2 + 1/2 + log(pi)/2'
```
It should not be necessary to use strings for comparison if you write `S(1)/2` instead of `1/2`.

czgdp1807 replied:
Thanks for the suggestion. I have removed the string comparison.

jksuom commented on sympy/stats/tests/test_rv.py:
```diff
@@ -107,6 +107,14 @@ def test_IndependentProductPSpace():
+    D = Die('D', sides = 4)
+    G = Geometric('G', 0.5)
+    assert str(H(X, X > 0)) == '-log(2)/2 + 1/2 + log(pi)/2'
+    assert simplify(H(D, D > 2)) == log(2)
```
Why `simplify`?

czgdp1807 replied:
I have removed it. The tests are passing without it on my system. I just used it for safety. Thanks for asking. It wasn't required here.

czgdp1807 committed:
removed unused simplify

czgdp1807 committed:
removed unused simplify

czgdp1807 committed:
removed string comparison

czgdp1807 commented:
@jksuom I have made the necessary changes and all checks are passing. Please suggest more changes, if any?

jksuom commented:
I think this looks fairly good now. The entropy is computed as an expectation in a single function. I find it a little disturbing that a pdf given as a dict has to be treated separately. Maybe that could also be transformed to a random variable, though not necessarily in this PR.

@Upabjojr What do you think?

czgdp1807 commented:
@jksuom Infact that's disturbing for me too. I thought a lot for the unification of the entropy calculation logic for distributions of PDF as a dict. I tried to convert dict into Piecewise. If there exists any method/function which can convert dict into Piecewise then the whole logic can be unified.

jksuom commented:
I think this is good to go. Thanks.
